### PR TITLE
*: support "show create user"

### PR DESCRIPTION
**Pull request by lnhote**
### What problem does this PR solve? 
Support SHOW CREATE USER as in MySQL 5.7.
fix https://github.com/pingcap/tidb/issues/7733 

### What is changed and how it works?
1. add GetEncodedPassword function to Manager interface.
2. add support for ShowCreateTable in buildShowSchema function.
3. add support for ShowCreateUser in ShowExec.fetchAll function

### Check List 

Tests 

 - Unit test

Code changes

 - exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

### CODE REVIEW

**Comment by morgo** [comment]
@lnhote Thank you for working on this!  It looks like it just needs `go fmt`  to be run:
```
go fmt privilege/privileges/privileges.go
```

**Comment by lnhote** [comment]
Thanks for the help. Please take another look.

**Comment by morgo** [comment]
@lnhote can you please add tests for:
- a user that doesn't exist
- the user exists but the host portion doesn't match

I am getting the following behavior on a fresh install. Both are different results from MySQL:
```
mysql> SET PASSWORD='abcd';
2019/02/10 02:35:19.953 session.go:1639: [info] [CRUCIAL OPERATION] con:1 schema_ver:15 set password for user <nil> (by root@127.0.0.1).
Query OK, 0 rows affected (0.00 sec)

mysql> SHOW CREATE USER 'root'@'%'; <-- correct
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| CREATE USER for root@%                                                                                                                                            |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| CREATE USER 'root'@'%' IDENTIFIED WITH 'mysql_native_password' AS '*A154C52565E9E7F94BFC08A1FE702624ED8EFFDA' REQUIRE NONE PASSWORD EXPIRE DEFAULT ACCOUNT UNLOCK |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> SHOW CREATE USER 'root'@'nowhere'; <--- incorrect
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| CREATE USER for root@nowhere                                                                                                                                            |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| CREATE USER 'root'@'nowhere' IDENTIFIED WITH 'mysql_native_password' AS '*A154C52565E9E7F94BFC08A1FE702624ED8EFFDA' REQUIRE NONE PASSWORD EXPIRE DEFAULT ACCOUNT UNLOCK |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> show create user 'aaa'@'aaa'; <-- incorrect
2019/02/10 02:35:39.737 privileges.go:65: [error] Get user privilege record fail: user aaa, host aaa
+---------------------------------------------------------------------------------------------------------------------------+
| CREATE USER for aaa@aaa                                                                                                   |
+---------------------------------------------------------------------------------------------------------------------------+
| CREATE USER 'aaa'@'aaa' IDENTIFIED WITH 'mysql_native_password' AS '' REQUIRE NONE PASSWORD EXPIRE DEFAULT ACCOUNT UNLOCK |
+---------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by morgo** [comment]
Expected behavior:
```
mysql [localhost:5724] {msandbox} ((none)) > show grants for 'root'@'nowhere';
ERROR 1141 (42000): There is no such grant defined for user 'root' on host 'nowhere'
mysql [localhost:5724] {msandbox} ((none)) > show grants for 'aaa'@'aaa';
ERROR 1141 (42000): There is no such grant defined for user 'aaa' on host 'aaa'
```

File: executor/builder.go [hunk 1]
```diff
@@ -525,6 +525,9 @@ func (b *executorBuilder) buildShow(v *plannercore.Show) Executor {
 	if e.Tp == ast.ShowGrants && e.User == nil {
 		e.User = e.ctx.GetSessionVars().User
 	}
+	if e.Tp == ast.ShowCreateUser && e.User == nil {
+		e.User = e.ctx.GetSessionVars().User
+	}
```
Comment (morgo) [inline executor/builder.go]: For `show grants`, you can execute this command to see your own grants.  For `show create user` though, this syntax is not supported.  Is this required here?

File: executor/show.go [hunk 2]
```diff
@@ -843,6 +845,29 @@ func (e *ShowExec) fetchShowCollation() error {
...
+		return errors.New("miss privilege checker")
+	}
+	sql := fmt.Sprintf(`SELECT * FROM %s.%s WHERE User='%s' AND Host='%s';`,
+		mysql.SystemDB, mysql.UserTable, e.User.Username, e.User.Hostname)
+	rows, _, err := e.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(e.ctx, sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(rows) == 0 {
+		return ErrCannotUser.GenWithStackByArgs("SHOW CREATE USER",
```
Comment (morgo) [inline executor/show.go]: This is the wrong error code for MySQL compatibility.  It should be:
```
return ErrNonexistingGrant.GenWithStackByArgs(e.User.Username, e.User.Hostname)
```
You will need to add this error to `./executor/errors.go`:
```
@@ -48,6 +48,7 @@ var (
        ErrDBaccessDenied              = terror.ClassExecutor.New(mysql.ErrDBaccessDenied, mysql.MySQLErrName[mysql.ErrDBaccessDenied])
        ErrTableaccessDenied           = terror.ClassExecutor.New(mysql.ErrTableaccessDenied, mysql.MySQLErrName[mysql.ErrTableaccessDenied])
        ErrBadDB                       = terror.ClassExecutor.New(mysql.ErrBadDB, mysql.MySQLErrName[mysql.ErrBadDB])
+       ErrNonexistingGrant            = terror.ClassExecutor.New(mysql.ErrNonexistingGrant, mysql.MySQLErrName[mysql.ErrNonexistingGrant])
 )
 
 func init() {
@@ -63,6 +64,7 @@ func init() {
                mysql.ErrDBaccessDenied:              mysql.ErrDBaccessDenied,
                mysql.ErrTableaccessDenied:           mysql.ErrTableaccessDenied,
                mysql.ErrBadDB:                       mysql.ErrBadDB,
+               mysql.ErrNonexistingGrant:            mysql.ErrNonexistingGrant,
        }
        terror.ErrClassToMySQLCodes[terror.ClassExecutor] = tableMySQLErrCodes
 }
```

File: executor/builder.go (hunk 1 above)
Comment (morgo) [reply executor/builder.go]: (I checked out the code, and it should be safe to just remove these 3 lines).

**Comment by morgo** [comment]
@lnhote I left two review comments.  It looks good otherwise.  Thanks!

File: executor/show.go (hunk 2 above)
Comment (lnhote) [reply executor/show.go]: This is what I see in mysql of version 5.7.17. What is your mysql version?

```
mysql: [Warning] Using a password on the command line interface can be insecure.
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 6
Server version: 5.7.17 Homebrew

Copyright (c) 2000, 2016, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> show create user 'aaa'@'aaa';
ERROR 1396 (HY000): Operation SHOW CREATE USER failed for 'aaa'@'aaa'

```

Comment (morgo) [reply executor/show.go]: @lnhote I checked 5.7 again, you are correct :-)  Sorry about that.

File: executor/builder.go (hunk 1 above)
Comment (lnhote) [reply executor/builder.go]: I have removed these lines.

**Comment by morgo** [comment]
LGTM

**Comment by morgo** [comment]
@lnhote please resolve conflicts.

**Comment by lnhote** [comment]
@morgo PTAL.

**Comment by morgo** [comment]
LGTM
@jackysp PTAL

File: executor/show_test.go [hunk 3]
```diff
@@ -16,13 +16,13 @@ package executor_test
 import (
 	"context"
 	"fmt"
-
```
Comment (jackysp) [inline executor/show_test.go]: Please do not remove this line, keep the sys libs separate with the 3rd party libs.

Comment (lnhote) [reply executor/show_test.go]: Adding the blank line will cause check_dev to fail.

Comment (jackysp) [reply executor/show_test.go]: Why check_dev could pass in master branch? I think you've added a blank line with some white spaces, try to keep it as it used to be.

Comment (lnhote) [reply executor/show_test.go]: I fixed it.

**Pull request merged**